File: solved/shiritori.py
import sys

n = int(sys.stdin.readline())
seen = set()

def other_player(player):
    if player == 1: return 2
    elif player == 2: return 1
player_to_win = 1
winner = 0

prev = sys.stdin.readline()
for i in range(1, n):
    seen.add(prev)
    current = sys.stdin.readline()
    
    if current in seen: 
        winner = player_to_win
        break
    elif prev[-2] != current[0]:
        winner = player_to_win
        break
    else:
        player_to_win = other_player(player_to_win)
        prev = current

if winner == 0:
    print("Fair Game")
elif winner == 1:
    print("Player 2 lost")
elif winner == 2:
    print("Player 1 lost")

File: solved/shiritori.hs
import Data.Set
import Data.Typeable

read_n_lines 1 = do
    line <- getLine
    print (line ++ " " ++ line)
    return [line]

read_n_lines n = do
    l <- read_n_lines 1 
    ls <- read_n_lines (n-1)
    return (l ++ ls) 

other_player 1 = 2
other_player 2 = 1
other_player x = 0

game :: [String] -> Set String -> Int -> Int
game [] s p = 0

game (l:[]) s p = 
    if member l s then other_player p
    else 0

game (l1:l2:ls) s p = 
    if last l1 /= head l2 then p
    else if member l2 s then p
    else game (l2:ls) (insert l1 s) (other_player p)

main = do
    n <- readLn :: IO Int
    lines <- read_n_lines n
    let game_result = game lines empty 2
    if game_result == 1 then putStrLn "Player 1 lost"
    else if game_result == 2 then putStrLn "Player 2 lost"
    else putStrLn "Fair Game"